### ISSUE TITLE
OOP WinRT component for operations

### ISSUE THREAD

**Issue by hez2010** [files-community/Files#13632]
### What feature or improvement do you think would benefit Files?

We initially chose Win32 Shell APIs for file operations because we wanted the file operations to continue even after Files was closed, but Win32 Shell APIs are designed for automation, not for the general use by a file explorer. And it comes with its own limitations (e.g. slow, no built-in transfer speed calculation and etc.).

We should consider implementing file operations as an out-of-process WinRT component, so that we no longer need to rely on the Win32 Shell APIs, neither facing the performance and maintainability issues due to serialization. By using WinRT we can expose the APIs with WinRT types directly (such as `IAsyncOperation<T>` and `Stream`), and call them from another process.

### Requirements

It was possible only with C++/WinRT, and CsWinRT still doesn't have built-in support for authoring OOP WinRT exe server for now.
But recently a project [Shmuelie.WinRTServer](https://github.com/shmuelie/Shmuelie.WinRTServer) has this implemented, so I think we can start to investigate it. 

### Files Version

Planning for post 3.0

### Windows Version

10.0.22631.2506

**Comment by gave92** [comment]
You mean implementing file ops using Win32 API? E.g CopyFileEx, MoveFileWithProgress, etc?

**Comment by hez2010** [comment]
Yes. Implementing file ops using Win32 API in an external process and expose those APIs through WinRT ABI in an external process. This is similar with what we did before in fullTrustedProcess but not the same. By authoring a WinRT exe server, we don't need to serialize and dispath messages by ourselves, instead we cancan avoid the overhead of json serialization

**Comment by gave92** [comment]
Can a WinRT exe server run as administrator?

**Comment by hez2010** [comment]
> Can a WinRT exe server run as administrator? (without the main app running as admin)

We may not be able to activate an elevated WinRT type from a non-elevated process.

**Comment by yaira2** [comment]
Will this help us support third party handlers? #7518

**Comment by hez2010** [comment]
The investigation and experiment have done. Now I'm going to commit the implementation.
Note this is not a Code Quality work but a Feature work (which allows us to have a standalone background server for long-running tasks)

**Comment by JaiganeshKumaran** [comment]
Why not simply just run the regular executable but without a window open? It will be much simpler. There are not that many resources about creating an out-of-process WinRT server.

**Comment by hez2010** [comment]
> Why not simply just run the regular executable but without a window open? It will be much simpler. There are not that many resources about creating an out-of-process WinRT server.

Because then you will need to handle all serialization stuff and message dispatch by yourself, which results in major maintainability regression.

**Comment by JaiganeshKumaran** [comment]
> Because then you will need to handle all serialization stuff and message dispatching by yourself,
No I mean just do it in the same process as the main app. Just hide the window.

**Comment by hez2010** [comment]
> No I mean just do it in the same process as the main app. Just hide the window.

This will still face the concurrency issue. We want all operations to happen in a single process to allow us to update to LiteDB 5.

**Comment by yaira2** [comment]
> The investigation and experiment have done. Now I'm going to commit to the implementation. Note this is not a Code Quality work but a Feature work (which allows us to have a standalone background server for long-running tasks)

That's good to hear, keep us updated!

**Comment by JaiganeshKumaran** [comment]
> We want all operations to happen in a single process
I did mean the same process as the app, meaning hitting X on the window will only hide it and the operations will continue to run.

**Comment by 0x5bfa** [comment]
> This will still face the concurrency issue. We want all operations to happen in a single process to allow us to update to LiteDB 5.

What if we are going to make the app single instantiated? When user tries to open another window, Program class can redirect and App class can create another Windows class within the same App class while it would make things much complicated such as the DI container and MainWindow.Instance property. Tracked in #12689. In this case, oop WinRT server wouldn't be required.

Speaking of this, how we can cope with #7275? IFileOperation runs on STA. I wonder how File Explorer managed to run in parallel.

@hez2010 What do you think?
